src/wodan/wodan.ml

```diff
@@ -1663,50 +1614,21 @@ struct
               List.iter
                 (fun key -> KeyedMap.remove entry.logdata.contents key)
                 !to_remove;
-              (entry.logdata).value_end
+              entry.logdata.value_end
               <- entry.logdata.value_end - !remove_size;
-              let clo_out = ref block_end in
-              let clo_out1 = ref block_end in
-              let clo = ref (block_end - childlink_size) in
-              let children = Lazy.force entry.children in
-              let children1 = KeyedMap.create () in
-              let fc1 = KeyedMap.create () in
-              while !clo >= entry.childlinks_offset do
-                (* Move children data *)
-                let key1 =
-                  Cstruct.to_string
-                    (Cstruct.sub entry.raw_node !clo P.key_size)
-                in
-                if String.compare key1 median <= 0 then (
-                  clo_out1 := !clo_out1 - childlink_size;
-                  Cstruct.blit entry.raw_node !clo entry1.raw_node !clo_out1
-                    childlink_size;
-                  ( match KeyedMap.find_opt entry.children_alloc_ids key1 with
-                  | Some alloc_id ->
-                      KeyedMap.remove entry.children_alloc_ids key1;
-                      KeyedMap.xadd entry1.children_alloc_ids key1 alloc_id
-                  | None ->
-                      () );
-                  KeyedMap.remove children key1;
-                  KeyedMap.xadd children1 key1 (Int64.of_int !clo_out1);
-                  if KeyedMap.mem di key1 then (
-                    let child_alloc_id = KeyedMap.find di key1 in
-                    KeyedMap.remove di key1;
-                    KeyedMap.xadd fc1 key1 child_alloc_id ) )
-                else (
-                  clo_out := !clo_out - childlink_size;
-                  Cstruct.blit entry.raw_node !clo entry.raw_node !clo_out
-                    childlink_size;
-                  let key1 =
-                    Cstruct.to_string
-                      (Cstruct.sub entry.raw_node !clo_out P.key_size)
-                  in
-                  KeyedMap.replace_existing children key1
-                    (Int64.of_int !clo_out) )
-              done;
-              entry.childlinks_offset <- !clo_out;
-              entry1.childlinks_offset <- !clo_out1;
-              entry1.children <- Lazy.from_val children1;
+              let children1 =
+                KeyedMap.split_off_after entry.children median
+              in
+              KeyedMap.swap children1 entry.children;
+              let ca1 =
+                KeyedMap.split_off_after entry.children_alloc_ids median
+              in
+              KeyedMap.swap ca1 entry.children_alloc_ids;
+              entry1.children <- children1;
+              entry1.children_alloc_ids <- ca1;
```

Comment:
This seems dangerous to me, we might believe `entry1.children_alloc_ids` is taking the values of `ca1` as computed line 1624, while this is not the case due to the swap.
The same applies to `entry1.children`.
Maybe we should add a comment here.

More generally, I find this `swap` function a bit confusing, maybe we should think of a better way to achieve affectations.
I dont think adding `mutable`s is a good solution because this means double mutability, but we can probably turn our KeyedMap into a functional datastructure and then use `mutable`s here.
Telle me what you think.

Reply:
The point of swap here is not to perform affectations, just to provide a way of splitting before or at a value (the swaps all follow a split_off_after call). There's more than one way to write it, sure. I can put that inside the Keyedmap module instead.
I'd rather have Keyedmap stay imperative, because it's performance critical and we may yet find or write a faster implementation.

Reply:
Added a function to split off before or at a value, reducing the use of swap.